compiler/cpp/src/generate/t_rs_generator.cc

```diff
@@ -306,823 +252,164 @@ void t_rs_generator::generate_enum(t_enum* tenum) {
-          indent_up();
-            indent(f_mod_) << "try!(iprot.read_field_end(transport));\n";
-            indent(f_mod_) << "break;\n";
-          indent_down();
-          indent(f_mod_) << "}\n";
-
-          vector<t_field*>::const_iterator m_iter;
-          const vector<t_field*>& members = tstruct->get_members();
-          for (m_iter = members.begin(); m_iter != members.end(); ++m_iter) {
-            generate_field_read(*m_iter);
-          }
-
-          indent(f_mod_) << "(_, ftype, _) => {\n";
-          indent_up();
-            indent(f_mod_) << "try!(iprot.skip(transport, ftype));\n";
-          indent_down();
-          indent(f_mod_) << "}\n";
-
+        generate_service_method_arglist(tfunction->get_arglist()->get_members());
         indent_down();
-        indent(f_mod_) << "}\n";
-        indent(f_mod_) << "try!(iprot.read_field_end(transport));\n";
-      indent_down();
-      indent(f_mod_) << "}\n";
-      indent(f_mod_) << "try!(iprot.read_struct_end(transport));\n";
-      indent(f_mod_) << "if have_result { Ok(()) } else { Err(ThriftErr::from(Error::ProtocolViolation)) }\n";
-    indent_down();
-    indent(f_mod_) << "}\n";
-  indent_down();
-  indent(f_mod_) << "}\n\n";
-}
-
-void t_rs_generator::generate_field_read(t_field* field) {
-  string qualified_name = "self." + to_field_name(field->get_name());
-  t_type* type = get_true_type(field->get_type());
-  bool is_optional = field->get_req() == t_field::T_OPTIONAL;
-
-  if (is_optional) {
-    indent(f_mod_) << "/* FIXME\n";
-  }
-
-  indent(f_mod_) << "(_, Type::" << render_protocol_type(type)
-                 << ", " << field->get_key() << ") => {\n";
-  indent_up();
 
-  generate_read_value(type, qualified_name, is_optional);
+        indent(f_mod_) << ") -> " << render_rs_type(tfunction->get_returntype()) << " => [\n";
 
-  indent(f_mod_) << "have_result = true;\n";
-  indent_down();
-  indent(f_mod_) << "}\n";
+        indent_up();
+        generate_service_method_arglist(tfunction->get_xceptions()->get_members());
+        indent_down();
 
-  if (is_optional) {
-    // FIXME
-    indent(f_mod_) << "*/\n";
-  }
+        indent(f_mod_) << "],\n";
+    }
 }
 
-void t_rs_generator::generate_read_value(t_type* type, const string& name, bool is_optional,
-                                         bool is_decl) {
-  string decl = is_decl ? "let mut " : "";
-  string prefix = is_optional ? "Some(" : "";
-  string suffix = is_optional ? ")" : "";
+void t_rs_generator::generate_service_generics(t_service* tservice) {
+  t_service* parent = tservice;
+  char generic = 'A';
 
-  if (type->is_base_type()) {
-    indent(f_mod_) << decl << name << " = " << prefix
-                   << "try!(iprot.read_" << render_suffix(type) << "(transport))"
-                   << suffix << ";\n";
+  while (parent && generic <= 'Z') {
+    f_mod_ << generic << ": " << parent->get_name() << ", ";
+    parent = parent->get_extends();
+    generic++;
```

Comment:
Minor: I have a habit of using the prefix form, as it does not have side effects.